### PR TITLE
[RFC] feat(nitro): embed observability dashboard in-process at /_workflow

### PR DESCRIPTION
**Pull request by pranaygp**
## Context

Using Workflow with a Nitro-based framework (Nitro v2/v3, Nuxt) used to run the local observability UI as a *separate process*: the dev-only `/_workflow` handler booted the entire `@workflow/web` Express server on a random port and **302-redirected** to it. That second server, second port, and cross-origin redirect is what this PR eliminates.

Now the dashboard is served **in-process, on the same origin/port**, at a configurable route (default `/_workflow`) — no second server, no redirect, no orphan process.

This is the first half of making `@workflow/nitro` an all-in-one integration. The other half — auto-starting the world on boot — is a **follow-up PR** that builds on #2544 (`ensureWorldStarted`) and is intentionally not included here.

## Changes

- **`@workflow/web`** — new framework-neutral `@workflow/web/handler` export (`createWorkflowWebHandler({ basename })`) that serves SSR + static client assets + RPC as a single Web `Request`→`Response` handler under a runtime mount path. The React Router build (Vite base `/`) is reprefixed at runtime — asset-manifest URLs + `publicPath` — so the dashboard is **self-contained under its mount** (`<basename>/assets/...`), never touching the host app's root namespace. Adds `@workflow/web/registry` (best-effort discovery) and makes the RPC/stream client basename-aware. The standalone `@workflow/web/server` (CLI at `/`) is unchanged.
- **`@workflow/nitro`** — `addDashboardHandler` now mounts that handler in-process (Nitro v2 `h3.fromWebHandler` + v3 native paths) instead of redirecting. New `dashboard?: boolean | { enabled?, path? }` option; **default = on in dev, off in prod**. When disabled nothing is registered, so production bundles carry no `@workflow/web` import (zero bundle/startup cost). Excluded on Vercel deploys.
- **`@workflow/cli`** — `workflow web` / `inspect --web` detect an already-running embedded dashboard (via the registry, health-checked) and open it instead of spawning a redundant server on `:3456`. `--standalone` forces the standalone UI.

Because the embedded data layer shares the host process's `process.env`, it reads the **same world** as the running app automatically (no extra wiring).

## Test plan

Verified manually across all three targets + prod gating + CLI:

| Check | Result |
|---|---|
| `workbench/nitro-v3` (v3 native) `/_workflow` | 200 HTML, assets reprefixed to `/_workflow/assets/…` + served `text/javascript`, RPC `success:true` |
| `workbench/nitro-v2` (v2 h3) | 200, UI + assets + RPC all working |
| `workbench/nuxt` (Vite SSR) | 200, UI + assets; deep-link `/_workflow/run/x` served by the handler (no SPA-fallback issue) |
| No second port / orphan | only the app's port listens; registry isolates per-pid |
| Prod build (default) | `/_workflow` serves the app's own `index.html`; no `@workflow/web` in the bundle |
| CLI coordination | detects embedded URL; `--standalone` bypasses |
| Automated | `@workflow/web` + `@workflow/nitro` unit tests pass; Biome clean |

## Follow-up (separate PR, not here)

Auto-start the world on server boot via a Nitro plugin (`ensureWorldStarted`), gated off Vercel — depends on #2544 merging first.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/nitro/src/index.ts [hunk 1]
```diff
@@ -210,8 +210,21 @@ export default {
...
+      // and node_modules import don't survive a serverless bundle — users use
+      // the hosted Vercel dashboard there).
+      const dashboardOption = nitro.options.workflow?.dashboard;
+      const dashboardEnabled =
+        typeof dashboardOption === 'object'
+          ? (dashboardOption.enabled ?? nitro.options.dev)
+          : (dashboardOption ?? nitro.options.dev);
+      const dashboardPath =
+        (typeof dashboardOption === 'object' && dashboardOption.path) ||
+        '/_workflow';
```
Comment (RihanArfan) [inline packages/nitro/src/index.ts]: `dashboardPath` is used raw both for Nitro route registration (`[basename, basename + '/**']`) and passed through as the handler `basename`, but normalization (`normalizeBasename`) only happens inside the handler. So a custom `path` can make the two disagree:

- `'/_workflow/'` (trailing slash) -> routes `/_workflow/` and `/_workflow//**`, handler normalizes to `/_workflow`
- `'admin/wf'` (no leading slash) -> Nitro routes don't start with `/`, and `reprefixBuild` emits non-root-absolute asset URLs
- `'/'` (root) -> routes `/` and `//**`; the `/` catch-all would swallow the whole host app

Default is fine, but since `path` is documented as configurable, suggest normalizing once here (force leading slash, strip trailing, disallow root) before both the route registration and the handler get it.

File: packages/nitro/src/index.ts [hunk 2]
```diff
@@ -294,70 +307,85 @@ export default {
...
+          handlerPromise = null;
           throw error;
         });
       }
-      return serverPromise;
+      return handlerPromise;
     }
   `;
 
   if (!nitro.routing) {
```
Comment (RihanArfan) [inline packages/nitro/src/index.ts]: We have `isNitroV2()` to centralize v2/v3 detection (it prefers `nitro.meta.majorVersion` and only falls back to `!nitro.routing`), but this branches on `!nitro.routing` directly. Equivalent in practice, but worth using the helper for consistency. Pre-existing across the other handlers too, so could be a follow-up sweep rather than blocking here.

File: packages/nitro/src/index.ts (hunk 1 above)
Comment (pranaygp) [reply packages/nitro/src/index.ts]: Done in `c235d1a4` — added a `normalizeDashboardPath()` that runs once on the configured `path` *before* both the route registration (`[path, path + '/**']`) and the handler `basename` receive it. It forces a single leading slash, strips trailing slashes, and falls back to the default when the result is root `/`. So all three cases you listed now resolve consistently: `/_workflow/` → `/_workflow`, `admin/wf` → `/admin/wf`, and `/` → the default `/_workflow` (no `/**` catch-all swallowing the host app).

File: packages/nitro/src/index.ts (hunk 2 above)
Comment (pranaygp) [reply packages/nitro/src/index.ts]: Done in `c235d1a4` — swept all the handler-level `!nitro.routing` checks (`addDashboardHandler`, `addVirtualHandler`, `addManifestHandler`, `writeManifestHandler`) over to `isNitroV2(nitro)` for consistent v2/v3 detection. The only remaining `!nitro.routing` is the fallback inside `isNitroV2()` itself.

File: packages/nitro/src/types.ts [hunk 3]
```diff
@@ -35,6 +35,23 @@ export interface ModuleOptions {
...
+   * - `{ enabled, path }` — toggle and/or change the mount path.
+   *
+   * Defaults to **on in dev, off in production builds**. When disabled the
+   * route is never registered, so production bundles carry no `@workflow/web`
+   * import and pay zero bundle-size or startup cost. Never mounted for Vercel
+   * deploys (use the hosted Vercel dashboard there).
+   *
+   * @default nitro.options.dev
+   */
+  dashboard?: boolean | { enabled?: boolean; path?: string };
```
Comment (VaguelySerious) [inline packages/nitro/src/types.ts:54]: AI Review: Note

This adds a user-facing `dashboard` option, but the PR touches no docs. Two updates needed in `docs/content/docs/v5/api-reference/workflow-nitro/index.mdx`:

1. The behavior line still says *"Serves a redirect to the local observability dashboard at `/_workflow` in development"* — now stale; it's served in-process with no redirect.
2. Add `dashboard` to the module-options table (type, default `nitro.options.dev`, configurable `path`).

Per repo convention, user-facing changes should be reflected in `docs`.

File: packages/nitro/src/index.ts [hunk 4]
```diff
@@ -346,71 +360,103 @@ export default {
...
+ * embedded handler's own basename normalization can't disagree.
+ *
+ * `dashboardPath` feeds two consumers: the Nitro route registration
+ * (`[path, path + '/**']`) and the handler's `basename` (which re-normalizes
+ * internally via `normalizeBasename`). Normalizing once, here, keeps them in
+ * sync. We force a single leading slash, strip trailing slashes, and reject
+ * the root mount (whose `/**` catch-all would swallow the host app), falling
+ * back to the default.
+ */
+function normalizeDashboardPath(path: string): string {
```
Comment (VaguelySerious) [inline packages/nitro/src/index.ts:376]: AI Review: Note

No automated coverage for the new logic. `normalizeDashboardPath` is a pure function that's cheap to unit-test and has real edge cases worth pinning: trailing slashes, duplicate leading slashes, and the root-mount rejection that falls back to the default (the `/**` catch-all swallowing the host app is exactly what this guards against). The handler's asset-reprefixing and the CLI registry-detection path are also untested.

I verified all of these manually (in-process serve, asset reprefix, custom `path` remount, `--standalone` bypass, embedded detection), but a couple of unit tests would lock the behavior in against regressions.

File: packages/cli/src/lib/inspect/web.ts [hunk 5]
```diff
@@ -10,6 +10,55 @@ export const getHostUrl = (webPort: number) => `http://localhost:${webPort}`;
...
+      }
+    }
+    try {
+      const res = await fetch(entry.url, {
+        method: 'HEAD',
+        redirect: 'manual',
+        signal: AbortSignal.timeout(1500),
+      });
+      // Any non-5xx response means something is serving that route.
+      if (res.status < 500) live.push(entry);
```
Comment (VaguelySerious) [inline packages/cli/src/lib/inspect/web.ts:54]: AI Review: Nit

`res.status < 500` treats a 404 at the recorded URL as a live dashboard. The pid-liveness check above handles the common stale case, so this is low-risk, but a process that's alive yet no longer serving that exact mount (e.g. the path was reconfigured) would lead the CLI to open a 404. Consider narrowing to `< 400` (a 404 means nothing is serving that route).

**Review by VaguelySerious — APPROVED** [review]
AI review: no blocking issues

**Validated locally** by building a fresh Nitro v3 app against this PR's tarball and exercising the embedded dashboard end-to-end:

| Check | Result |
|---|---|
| `/_workflow` served in-process (not the old 302 redirect) | 200 `text/html` |
| Client assets reprefixed under the mount, correct MIME | `/_workflow/assets/*.js` as `text/javascript`, no root-namespace leakage |
| RPC routed under the mount | `/_workflow/api/rpc` hits the handler, not the SPA fallback |
| Deep-link served by the handler | `/_workflow/run/<id>` → 200 (no SPA-fallback issue) |
| Single port / no orphan | only the app port listens; no second server spawned |
| Embedded handler self-registers | registry entry with url + basename + world + pid |
| `workflow web` defers to the embedded dashboard | detects + opens it, does not spawn a redundant server |
| `--standalone` | bypasses detection, spawns the standalone UI |
| Custom `dashboard: { path }` | remounts (assets follow), default path 404s |
| Live UI | runs render over the same origin |

Changes I'd like (all non-blocking) are inline: docs not updated for the new behavior/option, and the new logic lacks automated coverage. Approving.

**Review by RihanArfan — APPROVED** [review]
